### PR TITLE
Add netbox pillar

### PR DESCRIPTION
**Pull request by kris-steinhoff**
### What does this PR do?

Add a pillar module that adds data to the Pillar structure from a [NetBox](http://netbox.readthedocs.io/en/latest/) API. Optionally creates the `proxy` structure.

### Tests written?

No

### Commits signed with GPG?

No

### CODE REVIEW

File: salt/pillar/netbox.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+This module currently only supports the napalm proxy minion and assumes
+you will use SSH keys to authenticate to the network device.  If password
+authentication is desired, it is recommended to create another ``proxy``
+key in pillar_roots (or git_pillar) with just the ``passwd`` key and use
+:py:func:`salt.renderers.gpg <salt.renderers.gpg>` to encrypt the value.
+If any additional options for the proxy setup are needed they should also be
+configured in pillar_roots.
+'''
+
```
Comment (terminalmage) [inline salt/pillar/netbox.py]: Please add the following here:

```python
from __future__ import absolute_import, print_function, unicode_literals
```

**Review by terminalmage — CHANGES_REQUESTED** [review]
We need some imports from `__future__` here. Also, please update the logging here so that you pass the args to the log handler instead of performing string replacement manually? Our logging handler normalizes str/unicode mismatches on Python 2, so you should just do your logging like this:

```python
log.debug('foo = %s', foo)
```

Doing the logging the way you have it now will result in the string replacement happening before our logging handler has a chance to normalize any str/unicode mismatches.

**Comment by ksofa2** [comment]
@terminalmage Thanks. We've made the requested changes. Also squashed the commits to clean up a mess I made in merging. (Sorry about that.)

File: salt/pillar/netbox.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        log.warn('API query failed for "%s", status code: %d',
+            minion_id, device_results.status_code)
+
+    # Assign results from API call to "netbox" key
+    try:
+        devices = device_results.json()['results']
+        if len(devices) == 1:
+            ret['netbox'] = devices[0]
+        elif len(devices) > 1:
+            log.error('More than one device found for "%s"' % minion_id)
```
Comment (terminalmage) [inline salt/pillar/netbox.py]: This log entry still needs to be changed per my earlier comments.

**Review by terminalmage — CHANGES_REQUESTED** [review]

**Comment by ksofa2** [comment]
Fixed.

File: salt/pillar/netbox.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            if platform_results.status_code != requests.codes.ok:
+                log.info('API query failed for "%s", status code: %d',
+                         minion_id, platform_results.status_code)
+
+            # Assign results from API call to "proxy" key
+            ret['proxy'] = {
+                'host': str(ipaddress.IPv4Interface(
+                            ret['netbox']['primary_ip4']['address']).ip),
+                'driver': platform_results.json()['napalm_driver'],
+                'proxytype': 'napalm',
```
Comment (mirceaulinic) [inline salt/pillar/netbox.py]: This should not be hardcoded like that. Maybe at least have it as an option or something.

File: salt/pillar/netbox.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+A module that adds data to the Pillar structure from a NetBox API.
+
+
+Configuring the NetBox ext_pillar
+====================================
+
+.. code-block:: yaml
+
+  ext_pillar:
+    - netbox:
```
Comment (mirceaulinic) [inline salt/pillar/netbox.py]: I would argue that this Pillar module should not be named as `netbox`, as its scope is very, very specific to retrieving data from NetBox and populating the `proxy` key into the Pillar. While this is very useful and nice, maybe we should emphasise the scope and rename it, say `netbox_proxy` or something similar.

To clarify: a `netbox` external Pillar I would expect to provide me _all_ the NetBox data related to that device (i.e, from the  `dcim/devices/<device ID>/` endpoint).

**Review by mirceaulinic — CHANGES_REQUESTED** [review]
Nice work @ksofa2. I think we should rename this module appropriate to its scope, and document what it does exactly (the docstring doesn't currently reflect that).

File: salt/pillar/netbox.py (hunk 4 above)
Comment (ksofa2) [reply salt/pillar/netbox.py]: It's creating the `proxy` key if requested, but this isn't just doing that. It also is returning the device info in the `netbox` key. It's a raw dump of what it gets back from the API call.

Comment (mirceaulinic) [reply salt/pillar/netbox.py]: Ah, okay, I misread that - please disregard.

File: salt/pillar/netbox.py (hunk 3 above)
Comment (ksofa2) [reply salt/pillar/netbox.py]: Good point, but given the structures we get back from NetBox, I'd like to leave it hardcoded and update the documentation so say that only napalm proxtypes are supported.

I added a code change to only produce the `proxy` key if the napalm_driver is defined in the device's platform entry.

File: salt/pillar/netbox.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def ext_pillar(minion_id, pillar, *args, **kwargs):
+    '''
+    Query NetBox API for minion data
+    '''
+
+    # Pull settings from kwargs
+    api_url = kwargs['api_url'].rstrip('/')
+    api_token = kwargs['api_token']
```
Comment (mirceaulinic) [inline salt/pillar/netbox.py]: I believe the token should not be mandatory, as this Pillar module only fetches data (the token is required only when writing data into NetBox) - see https://docs.saltstack.com/en/develop/ref/modules/all/salt.modules.netbox.html as an example.

**Review by mirceaulinic — CHANGES_REQUESTED** [review]

**Review by mirceaulinic — APPROVED** [review]
I wouldn't have any further comments on this, just to ask you to rebase and squash (but make sure to leave aside commits such as https://github.com/saltstack/salt/pull/46559/commits/ed6cdb775904e1efcfbdbbe95428e18be191f58f or https://github.com/saltstack/salt/pull/46559/commits/2c6d44ad51b777ab50765f4c1cd6aa04776752a2 which are unrelated to this addition). So I will defer to @terminalmage for more input.

Thanks for your contribution @ksofa2.

**Review by terminalmage — APPROVED** [review]
:+1:

**Comment by ksofa2** [comment]
Thanks @mirceaulinic and @terminalmage!

I've squashed and rebased onto develop. Please let me know if there's anything else I need to do.

**Pull request merged**